src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Combinators.hs

```diff
@@ -120,6 +121,11 @@ desugarRange i b
           | c == zero = ($ j) * (one - ($ k))
           | otherwise = one + ($ k) * (($ j) - one)
 
+safeZero :: (Arithmetic a, Traversable f) => ArithmeticCircuit a f -> ArithmeticCircuit a f
+safeZero r = circuitF $ do
+    is' <- runCircuit r
+    for is' $ \i -> newConstrained (\x j -> (x i - one) * x j) (\x -> x i - one)
```

Comment:
As written, $x_j = x_i - 1$, so the constraint simplifies to $(x_i - 1) x_j = x_j^2$ which can be simplified to just $x_j$ or `constraint (\x -> x i - one)` altogether without introducing new variable.
Is there a typo or am I misunderstanding something?